vm/templates/wallet/wallet.go

```diff
@@ -1,56 +1,211 @@
+	if err != nil {
+		return nil, fmt.Errorf("new wallet template: failed to load template account: %w", err)
+	} else if len(templateAccount.State) == 0 {
+		return nil, errors.New("new wallet template: template account state is empty")
+	}
+	templateCode := templateAccount.State
 
-//go:generate scalegen
+	// Load the wallet state
+	walletAccount, err := host.Get(host.Principal())
+	if err != nil {
+		return nil, fmt.Errorf("new wallet template: failed to load wallet principal account: %w", err)
+	} else if len(walletAccount.State) == 0 && !host.IsSpawn() {
+		// If this is a spawn we expect the current state to be empty
+		return nil, errors.New("new wallet template: wallet account state is empty for non-spawn")
+	} else if len(walletAccount.State) != 0 && host.IsSpawn() {
+		// If this is a spawn we expect the current state to be empty
+		return nil, errors.New("new wallet template: wallet account state is not empty for spawn")
+	}
+	walletState := walletAccount.State
+
+	return &Wallet{host, templateCode, walletState}, nil
+}
 
 // Wallet is a single-key wallet.
 type Wallet struct {
-	PublicKey core.PublicKey
+	host         core.Host
+	templateCode []byte
+	walletState  []byte
 }
 
 // MaxSpend returns amount specified in the SpendArguments for Spend method.
-func (s *Wallet) MaxSpend(args any) (uint64, error) {
-	// TODO(lane): rewrite to use the VM
-	// switch method {
-	// case core.MethodSpawn:
-	// 	return 0, nil
-	// case core.MethodSpend:
-	// 	return args.(*SpendArguments).Amount, nil
-	// default:
-	return 0, fmt.Errorf("%w: unknown method", core.ErrMalformed)
-	// }
+func (s *Wallet) MaxSpend(payload []byte) (uint64, error) {
+	maxgas := int64(s.host.MaxGas())
+	if maxgas < 0 {
+		return 0, errors.New("gas limit exceeds maximum int64 value")
+	}
+
+	var unmarshaled athcon.Payload
+	err := gossamerScale.Unmarshal(payload, &unmarshaled)
+	if err != nil {
+		return 0, fmt.Errorf("%w: malformed spawn payload", core.ErrMalformed)
+	}
+
+	// Check the method selector
+	// We define MaxSpend for any method other than spend to be zero for now.
+	spendSelector, _ := athcon.FromString("athexp_spend")
+	if unmarshaled.Selector == nil || *unmarshaled.Selector != spendSelector {
+		return 0, nil
+	}
+
+	// construct the payload. this requires some surgery to replace the method maxSpendSelector.
+	maxSpendSelector, _ := athcon.FromString("athexp_max_spend")
+	maxGasPayload := athcon.Payload{
+		Selector: &maxSpendSelector,
+		Input:    unmarshaled.Input,
+	}
+	maxGasPayloadEncoded, err := gossamerScale.Marshal(maxGasPayload)
+	if err != nil {
+		return 0, fmt.Errorf("marshaling maxSpend payload: %w", err)
+	}
+	executionPayload := athcon.EncodedExecutionPayload(s.walletState, maxGasPayloadEncoded)
+
+	// Instantiate the VM
+	// Use a mock host to ensure that no state changes occur.
+	host := s.host.Clone()
+	vmhost, err := vmhost.NewHost(host)
+	if err != nil {
+		return 0, fmt.Errorf("loading Athena VM: %w", err)
+	}
```

Comment:
Here and in few other places the VM is not destroyed (missing call to `Destroy`)
```suggestion
	vmhost, err := vmhost.NewHost(host)
	if err != nil {
		return 0, fmt.Errorf("loading Athena VM: %w", err)
	}
	defer vmhost.Destroy()
```

Reply:
good catch